common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,225 @@
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service for watching file at path: {}", this.pathOfFileToWatch);
+
+            String fileName = nameOfFileToWatch();
+            Path directoryPath = directoryOfFileToWatch();
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.pathOfFileToWatch);
+
+            while (true) {
+                watchKey = watchService.take();
+                log.info("Retrieved and removed watch key for watching file at path: {}", this.pathOfFileToWatch);
+
+                // Each file modification/create usually results in the WatcherService reporting the WatchEvent twice,
+                // as the file is updated twice: once for the content and once for the file modification time.
+                // These events occur in quick succession. We wait for 200 ms., here so that the events get
+                // de-duplicated - in other words only single event is processed.
+                //
+                // See https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-
+                // of-the-same-event for a discussion on this topic.
+                //
+                // If the two events are not raised within this duration, the callback will be invoked twice, which we
+                // assume is not a problem for applications of this object. In case the applications do care about
+                // being notified only once for each modification, they should use the FileModificationPollingMonitor
+                // instead.
+                Thread.sleep(200);
+
+                if (watchKey != null) {
+                    Optional<WatchEvent<?>> modificationDetectionEvent = watchKey.pollEvents()
+                            .stream()
+                            .filter( // we only care about changes to the specified file.
+                                    event -> event.context().toString().contains(fileName))
+                            .findAny(); // we only care to know whether the
+
+                    if (modificationDetectionEvent.isPresent()) {
+                        log.info("Detected that the file [{}] has modified", this.pathOfFileToWatch);
+                        callback.accept(modificationDetectionEvent.get());
+                    }
+
+                    boolean isKeyValid = watchKey.reset();
+                    log.debug("Done resetting watch key, so that it can receive further event notifications.");
```

Comment:
I asked you in a previous review of this PR to make your logs less verbose. There is no need to include all these explanations in here. Just indicate at which point in your processing you are at (the bare minimum).

Reply:
I've modified it to make the log less verbose as suggested.